packages/@uppy/provider-views/src/ItemList.js

```diff
@@ -27,21 +29,39 @@ module.exports = (props) => {
         // making <ul> not focusable for firefox
         tabindex="-1"
       >
-        {props.folders.map(folder =>
-          Item({
+        {props.folders.map(folder => {
+          return Item({
             ...getSharedProps(folder, props),
             type: 'folder',
             isDisabled: props.isChecked(folder) ? props.isChecked(folder).loading : false,
             handleFolderClick: () => props.handleFolderClick(folder)
           })
-        )}
-        {props.files.map(file =>
-          Item({
-            ...getSharedProps(file, props),
+        })}
+        {props.files.map(file => {
+          console.log(file)
+          const passesRestrictions = props.passesRestrictions(
+            remoteFileObjToLocal(file)
+          )
+          const sharedProps = getSharedProps(file, props)
+          let restrictionReason = passesRestrictions.reason
+          if (!props.canSelectMore && !sharedProps.isChecked) {
+            if (props.maxNumberOfFiles) {
+              restrictionReason = sharedProps.i18n('youCanOnlyUploadX', { smart_count: props.maxNumberOfFiles })
```

Comment:
If we have both restrictions enabled & `!props.canSelectMore && !sharedProps.isChecked` is true, then we'll always get `'youCanOnlyUploadX'` restrictionReason.

Example: getting `'You can only upload 80 files'` titles with `{ restrictions: { maxTotalFileSize: 291000, maxNumberOfFiles: 80 } }`
![image](https://user-images.githubusercontent.com/7578559/98650401-a65be080-235a-11eb-8919-e4e32c2efda6.png)

Reply:
In general looks like `.canSelectMore` makes it more confusing than necessary.

What do you think about returning the `files` argument to `_checkRestrictions(files, newFile)`, and passing `_checkRestrictions([...currentSelection, ...addedFiles], newFile)`?
We wouldn't need to repeat the `youCanOnlyUploadX` & `exceedsSize2` logic at all then.

Reply:
Totally agreed, your suggestion is much simpler, don’t know why I went the crazy way. Refactored! Thank you.